### ISSUE TITLE
Investigate the possibility of using koopman operator to approximate the motion model in the Kalman Filter

### ISSUE THREAD

**Issue by Zweisteine96** [ami-iit/element_physics-neural-inertial-odometry#7]
I spent some time on this idea, instead of using the normal linearization method around the current estimates (as usually used in the Kalman filter setting), I want to know if it's a good idea to use the Koopman operator theory to approximate a linearized motion model for EKF. 

There is a study that applied a data-driven Koopman operator and a ESEKF for the state estimation of a drone, see [here](https://vodafone-chair.org/pbls/ketong-zheng/Data-Driven_Koopman_Operator-Based_Error-State_Kalman_Filter_for_Enhanced_State_Estimation_of_Quadrotors_in_Agile_Flight.pdf). 

Another work that utilized only Koopman theory to estimate the robot centroidal state is [here](https://arxiv.org/pdf/2403.13366).

**Comment by Zweisteine96** [comment]
After several days exploration, from the basics of the Koopman theory to its applications for estimation and control, my impression is that there are both pros and cons of this idea, but maybe the cons overweight the pros.

### A potential framework for Koopman-Kalman-Estimator
#### Offline phase (Koopman model learning)
- Define a set of lifting functions/observables `g(x)` to obtain the lifted states `z=g(x)`.
- Use the collected data $(x_k, x_{k+1})$ (where $x$ is the original state vector) to generate pairs lifted states $(z_k, z_{k+1})$.
- Apply `Dynamic Mode Decomposition (DMD)` or `Extended DMD` or any other potential approach to find the best linear koopman operator `K` such that $z_{k+1} = Kz_k$. 

#### Online phase (Kalman filtering)
- Initialize the original states $\hat{x}_0$ and corresponding lifted states $\hat{z}_0$, state covariance matrix $\hat{P}_0$.
- At each time step `i`, do the followings:
  - **Prediction**: predict the next lifted state using the learned linear dynamics $\hat{z}_{k|k-1} = K * \hat{z}_{k-1|k-1}$. Predict the state covariance matrix $P_{k|k-1} = KP_{k-1|k-1}K^T + Q$  (where $Q$ is the system estimate noise covariance).
  - **Measurement model**: you'll need a measurement model `h(z)` that relates the lifted states `z` to a set of expected measurements $y_k$. You'll also need the measurement noise covariance `R`.
  - **Compute the Kalman gain**: based on the state covariance `P`, measurement model Jacobian `H` and measurement noise covariance `R`.
  - **Correction**: use the kalman filter (depends on the linearity of the measurement model, this can be KF or EKF or ESEKF and etc.) to update the estimates based on the prediction at first step, kalman gain, and measurements.
  - **Retrieval**: extract the original state estimate $\hat{x}_k$ from the updated lifted state estimate $\hat{z}_k$.

### Pros
From the presented workflow (there might be other paradigm), the biggest advantage of using the Koopman operator for a Kalman Filter is that the Koopman can provide a theoretically nice linear model if a good set of observables and an accurate operator are found. This can avoid the linearization errors inherent in EKF applied to the original nonlinear system. 

Another benifit might be that the Koopman operator (ideally) captures the dynamics over a large region of the state space covered by the training data, unlike EKF which linearizes around the current estimate.

### Challenges
But, there are also several significant challenges that need to be carefully considered (that's also why I said the cons might overweight the pros :( ).
- **How to choose the observables `g(x)`?** Using the neural network could be an option. But how to define the dimension of the lifted states `z`? Note that the dimension of `z` can directly impact the computational cost of the Kalman Filter.
- **How accurate is the linear operator `K`?** What we can do is to find the best linear fit in the lifted space using the data we have. It might still be an approximation. If significant non-linearity remains in the evolution of state `z`, then there is pointless to utilize this Koopman operator for a Kalman Filter.
- **What to do with the measurement model?** The choices of measurements is not unique for a Kalman Filter, some of which can still be highly non-linear. How to relate the measurements to the lifted states? This relation can be even more non-linear than the original ones, and what to do then?
- Some other issues also exist, for example, noise modeling in the lifted space? Covaraince matrices in the lifted space?

To conclude, this idea of combining Koopman and Kalman might sound appealing at a first glance, but it can be very challenging in both theoretical development and experimental validation. Therefore, I might not considering investigate further in this direction.

But, who knows, maybe someday I'll come back to this issue.

**Comment by Zweisteine96** [comment]
Closing
